man/ffpvocab.1.in
.de CW
. nop \s-2\f[C]\\$*\f[]\s+2
..
.de CODE
.in +0.5i
. nop \s-2\f[C]\\$*\f[R]\s+2
.in -0.5i
..
.TH ffpvocab 1 "[@]DATE[@]" "Version [@]VERSION[@]" "FFP PHYLOGENY"
.SH NAME
ffpvocab \- This program determines the number of features which have frequencies above a threshold value.
.SH SYNOPSIS
.BI "ffpvocab [" "OPTION" "] ... [" "FILE" "] ..."
.SH DESCRIPTION
.PP
Given no options, the default behavior of the program is to
print out the average number of features with frequencies greater than
two in all rows.  This value is defined as the number of vocabulary features.
FFPs will be read from standard input if no options are supplied and 
.B ffpvocab 
is called non-interactively (i.e. as part of a pipeline) or with a "-" in the 
argument list.  If a multi-row FFP is supplied then the default behavior is to
average the number of vocabulary features across all rows.
.SH OPTIONS
.TP
.BI "\-f " "INT" ", --freq-thresh=" "INT"
Threshold to count a feature.  The default value is 2, which means a feature 
must have a frequency greater than or equal to 2, in order to be counted.
.TP
.B \-h, --help
.PP
.SH EXAMPLES
To count the number of features that occur 3 or more times
in the ffp:
.PP
.CODE ffpry -l 5 file.fna | ffpvocab -f 3 
.PP
To calculate word usage for a range of lengths
use
.B ffpvprof.
.PP
.SH AUTHOR
This program was written by Gregory E. Sims.
.SH "REPORTING BUGS"
Report bugs to <gesims@lbl.gov>.
.SH COPYRIGHT
Copyright (C) [@]COPY[@] Gregory E. Sims
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ffpvprof(1),
.BR ffpre(1),
.BR ffpreprof(1)


NEWS
Version 3.19
	Thanks to Joemar Taggana, some incompatible utilities 
	were discovered in Mac OSX -- including incompatible
	getopt and mktemp.  Corrected confusion in tutorial 
        p. 3 regarding -m (Thanks to Erin Shellman).  Corrected
	spelling of ffpjsd option --hamann to --hamman.

Version 3.18
	Rearranged ungetc character tests to come after
	conversion of pipe to file.  Fixed memory leak in
	hash init().  Added support for large files (> 2GB) on 32
        bit machines.

Version 3.17
	Fixed bug in tempfile creation in function 
        convertPipeToFile().

Version 3.16
	Fixed bug in ffpry with multiple headers parsed as a
	single genome entry.

Version 3.15
	Fixed bug in ffprwn.

Version 3.14
	Added --disable-gui options to ./configure script. This
	disables the perl based gui, so that the Perl/Tk module
	isn't required for installation.

Version 3.13
	Changed row terminal output for all programs from '\t\n' 
	(tab newline) to '\n' (newline) only.  This makes output
	from previous version incompatible with 3.13 but the
	ffps can be easily converted using:

	sed 's/\t$//' ffp > ffp.tmp ; mv ffp.tmp ffp

Version 3.12
	Small bug fix for ffprwn for GNU compiler on 64 bit machines
	where promotion of unsigned int to long unsigned int is
	different.  Thanks to Adrienne Breland at the University
	of Nevada, Reno for finding this.

Version 3.10
	Fixed minor bug in ffpry for fasta records pseudomolecule 
	input ENDING in NNNNNNNN, or other invalid characters.

Version 3.09
	Fixed bug in ffpdf, related to initialized GROUP variable.
	
	Updated license

Version 3.08

	ffpgui

	Added instructions to install a functional version of the
	perl/tk module for using cygwin, in INSTALL.

	Added error messages in ffpgui for running a vocab or 
	relative entropy profile without loading a file first.

	Added additional sections to tutorial.pdf
       


Version 3.07
	Added signal handling to all executables.  
	Broken pipes will cause a termination error in a pipeline
	and will output 'PROGRAM: Broken pipe'

	Expanded ./doc/tutorial.in.
	Updated license.
	Fixed inappropriate default -f setting (changed from 1 to
	2) for ffpvprof.

Version 3.06
	Added ffptree utility, which creates neighbor joining and
	upgma trees.  Date and version automatically updated 
	during build process.
	
        Improved seeding of random generator when no seed is
	provide via command line option:  ffpry, ffpboot, ffpaa

	Added processing via ffptree to ffpgui.  The human 
	readable tree is displayed in the message window.

	Started tutorial file in ./doc directory. 

	Added tutorial.pdf to doc directory

Version 3.05
	Added length test since MAX_WORD_SIZE is 40: ffpry, ffpaa
	ffpre.

Version 3.04
	Added script ffpgui which is a Perl/Tk based GUI wrapper of
	the various utilities.

	Added --keys flag to ffpfilt to force keys to be output
	instead of values only.

Version 3.03
	Fixed overflow error with non-rabin karp hash used in ffpcol.

Version 3.02
	The ffpaat utility has been removed from this package.  The
	purpose of this utility was to count amino acid features in 
	all nine frames from DNA fasta input.  However	obtaining
	a proper translation of gene regions and using ffpaa is 
	generally more informative.  This utility can still be 
	compiled and used from previous versions, but will not longer
	be part of future releases. 

	Updated ffpre to Rabin-Karp hash, also fixed float overflow
	errors and cases of nan, or inf in output.  Text mode is now
	implemented in ffpre.

	Updated ffptxt to Rabin-Karp hash.


Version 3.01
	Implementation of Rabin-Karp hash in ffpaa utility.
	Similar speed gains as in ffpry obtained


	ffpaa -l 20 NC_008253.faa      

	ffpaa (v.3.0)                        ----------- 9.35 s 
	ffpaa (v.3.01)                       ----------- 0.15 s

	Input buffering options removed from both ffpaa and ffpry. The
	optimal buffering size for the system is determined automatically
	by using the stat.st_block size attribute of fstat or the compiler
	default, whichever is greater.  This is an effort to use the
	filesystem's most efficient block size.  
	
	Upcoming v3.02 will include Rabin-Karp implementation of ffpre.

Version 3.0
	Thanks to Nick Loman of University of Birmingham for pointing out
	the cryptic nature of the "key-value" error messages when both
	ffpfilt and ffpcol are used together in a pipeline successively.
	
	All error messages from the utilities state the program issuing
	the error.

	Revamped tmpfile creation for non-seekable piped input, to use
	most efficient buffering.

	Added extra unit tests to confirm piped vs file argument output
	is identical.

	ffpry changes:

	Beginning initial development of Rabin Karp Rolling hash function
	implementations which has replaced the hashing function used
	in version 2 in the ffpry executable.  This has provided a 
	significant speedup in k-mer counting.
	Here is a speed comparison:
	
	ffpry -l 10 -d NC_008253.fna             ------    9.7s (v.3.0)
        ffpry -l 10 -d NC_008253.fna             ------ 1m28.2s (v.2.30

	Removed all-possible/-a flag from ffpry.  The only way to generate
	columnar output from ffpry is via the -f flag and specify a specific
	feature set.

        -f specified features are now reverse complemented and classed as
	is the input sequence by default.	

	Added -r option which disables reverse complementing of the input
	sequence as well as operates on the feature list.  
	
	Fixed ffpry -f long option to correct value as per man file.
	
	Other:

	Fixed segmentation error in ffpboot

	Future directions:

	Rabin-Karp hash function will be implemented in ffpaa, ffptxt and
	ffpre, ffpaat
	
	

Version 2.30
	Fixed bug when encountering IUPAC codes in input files to ffpry.

Version 2.29
	Fixed bug when piping input into ffpmerge.

Version 2.28
	Addition of examples directory with trees from 
	Sims & Kim(2011) PNAS,108,8329-8334

        BUGS: ffpre doesn't report errors for trying to open

src/ffpvocab.c
/*****************************************************
* This code is distributed under a Non-commercial use 
* license.  For details see LICENSE.  Use of this
* code must be properly attributed to its author
* Gregory E. Sims provided that its use or derivative 
* use is non-commercial in nature.  Proper attribution        
* can be made by citing:
*
* Sims GE, et al (2009) Alignment-free genome 
* comparison with feature frequency profiles (FFP) and 
* optimal resolutions. Proc. Natl. Acad. Sci. USA.
* 106, 2677-82.
*
* Gregory E. Sims (C) 2010-2012
*
*****************************************************/
#include <stdio.h>
#include <getopt.h>
#include <stdlib.h>
#include <ctype.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include "utils.h"
#include "vstring.h"
#include "sighandle.h"
#include "../config.h"

char PROG_NAME[FILENAME_MAX];
#define VECTOR_SIZE 1000 /**< Initial guess for the number of columns in the FFP */
#define DEFAULT_THRESH 2 /**< The default frequency threshold for counting vocab features */

float vocab(FILE * fp, int threshold);

char usage_str[] = "Usage: %s [OPTION] ... [FILE] ...\n\
This program determines the number of features used in a vector\n\n\
Given no options, the defeault behavior of the program is to\n\
print out the number of features with frequencies greater than\n\
two in all rows.\n\
\t-f INT, --freq-thresh=INT\n\
\t-v, --version\n\
\t-h, --help\n\n\
Copyright (c) %s\n\
%s\n\
Contact %s\n";


int main(int argc, char **argv)
{
    FILE *fp;
    int opt;
    char fflag = 0;
    int threshold = DEFAULT_THRESH;
    int option_index = 0;

    static struct option long_options[] = {
	{"help", no_argument, 0, 'h'},
	{"version", no_argument, 0, 'v'},
	{"freq-thresh", required_argument, 0, 'f'},
	{"amino", no_argument, 0, 'a'},
	{"text", no_argument, 0, 't'},
	{0, 0, 0, 0}
    };

    initSignalHandlers();

    strcpy(PROG_NAME,basename( argv[0] ));

    while ((opt = getopt_long(argc, argv, "hf:atv",
			      long_options, &option_index)) != -1)

	switch (opt) {
	case 'f':
	    fflag = 1;
	    threshold = atoi(optarg);
	    break;
	case 'v':
	    printVersion();
	    exit(EXIT_SUCCESS);
	    break;
	case 'h':
	    printUsageStr();
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    printErrorUsageStr();
	    exit(EXIT_FAILURE);
	    break;
	}

//add code to detect what kind of ffp is being given as input.


    if (threshold < 1)
	fatal_msg("Feature threshold must be greater than or equal to 1\n");


    argv += optind;

    do {
	fp = stdin;
	if (*argv) {
	    if (!strcmp(*argv, "-"))
		fp = stdin;
	    else if ((fp = fopen(*argv, "r")) == NULL)
		fatal_msg("%s: %s.\n", *argv,strerror(errno));

	    if ( isDirectory(*argv) ) 
	    	fatal_msg("%s: %s\n",*argv, strerror(EISDIR));

	    argv++;
	} else if (isatty(STDIN_FILENO))
	    printErrorUsageStr();
    
	// check if not a seekable pipe
        if (!isRegularFile(fp)) 
	   fp = convertPipeToFile(fp);

	if (!isKeyBased(fp))
	   fatal_msg("%s: Not a key valued FFP.\n", *argv);

	printf("%e\t\n", vocab(fp, threshold));

	if (fp != stdin)
	    fclose(fp);

    } while (*argv);
    return EXIT_SUCCESS;
}



/**
 *
 * Determine the number of features in the FFP which
 * occur more than threshold number of times
 *
 * Determine the feature usage (vocabulary usage) of
 * the FFP stored in the File pointed to by fp.
 *
 * @param fp A file pointer
 * @param threshold A frequency threshold
 * @return The number of features above the frequency threshold
 *
 */



float vocab(FILE * fp, int threshold)
{
    int rows = 0;
    rows = 0;
    char ch[2];
    unsigned numFeature = 0;
    unsigned val;
    unsigned lineno=0;
    unsigned line_offset=0;
    int items;

    errno=0;
    while ( (items=fscanf(fp, "%*s %u%[\r\n]", &val,ch)) != EOF ) { 
	//fprintf(stderr,"%s %d\n",strerror(errno),items);
	if (errno) 
		fatal_msg("Parse error at line %u char %ld: %s.\n",
			lineno,ftell(fp)-line_offset, strerror(errno) );

	if (val >= threshold)
		numFeature++;
	
	switch (items) {
	        case 2:
			line_offset=ftell(fp);
			lineno++;
	      		rows++;		
		case 1:
		break;	
		default:
			fatal_msg("Parse error at line %u char %ld.\n",
			lineno,ftell(fp)-line_offset);
		break;
	}
    }


    if (ferror(fp)) 
    	fatal_msg("Read Error: %s",strerror(errno));

    return ((float) numFeature / rows);
}
